state-chain/pallets/cf-flip/src/tests.rs

```diff
@@ -649,11 +649,73 @@ mod test_tx_payments {
 }
 
 #[test]
-fn only_governance_can_set_slashing_rate() {
+fn update_pallet_config() {
 	new_test_ext().execute_with(|| {
+		// Only governance can update the pallet config
 		assert_noop!(
-			Flip::set_slashing_rate(RuntimeOrigin::signed(ALICE), Permill::from_percent(0)),
+			Flip::update_pallet_config(RuntimeOrigin::signed(ALICE), vec![].try_into().unwrap()),
 			sp_runtime::traits::BadOrigin,
 		);
+
+		// Slashing rate can be updated
+		let update_to_slashing_rate = Permill::from_percent(40);
+		assert_ne!(SlashingRate::<Test>::get(), update_to_slashing_rate);
+		assert_ok!(Flip::update_pallet_config(
+			RuntimeOrigin::root(),
+			vec![PalletConfigUpdate::SetSlashingRate(update_to_slashing_rate)]
+				.try_into()
+				.unwrap(),
+		));
+		assert_eq!(SlashingRate::<Test>::get(), update_to_slashing_rate);
+
+		// Fee scaling rate can be updated
+		let update_to_fee_scaling_rate =
+			FeeScalingRateConfig::DelayedExponential { threshold: 10, exponent: 2 };
+		// compare to fee we will update to with the fee before to see it updates
+		assert_ne!(FeeScalingRate::<Test>::get(), update_to_fee_scaling_rate);
+		assert_ok!(Flip::update_pallet_config(
+			RuntimeOrigin::root(),
+			vec![PalletConfigUpdate::SetFeeScalingRate(update_to_fee_scaling_rate)]
+				.try_into()
+				.unwrap(),
+		));
+		assert_eq!(FeeScalingRate::<Test>::get(), update_to_fee_scaling_rate);
 	});
 }
+
+mod test_fee_scaling_config {
+	use super::*;
+
+	#[test]
+	fn no_fee_scaling_does_not_scale_fee() {
+		for (pre_scaled_fee, call_count) in
+			[(0, 0), (1, 1), (u32::MAX, u16::MAX), (100u32, 1), (200, 9), (3, 4000)]
+		{
+			assert_eq!(
+				FeeScalingRateConfig::NoScaling.scale_fee(pre_scaled_fee, call_count),
+				pre_scaled_fee
+			);
+		}
+	}
+
+	#[test]
+	fn exponent_buffer_scales_fee() {
+		let pre_scaled_fee: u32 = 100;
+		const BUFFER: u16 = 10;
+
+		let large_buffer_config =
+			FeeScalingRateConfig::DelayedExponential { threshold: BUFFER, exponent: 2 };
+
+		// In reality the call count will be 1 (hence the +1), but we test 0 behaves reasonably in
+		// case.
+		for call_count in 0..=BUFFER {
+			assert_eq!(large_buffer_config.scale_fee(pre_scaled_fee, call_count), pre_scaled_fee);
+		}
+		for i in 0..BUFFER {
```

Comment:
kinda confusing having the buffer here, it's meaningless that we go to the buffer in this case, better to separate them - also as above, this test should catch the issue I mentioned

Reply:
Not sure if we need this test as well as the tests in `on_charge_transaction`

Reply:
Yeah, forgot to remove these.